core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java

```diff
@@ -19,263 +19,624 @@
+        @Override
+        public boolean supportsDiffableValues() {
+            return false;
+        }
+
+        @Override
+        public Diff<V> computeDiff(V value, V beforePart) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void writeDiffTo(Diff<V> value, StreamOutput out) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Diff<V> readDiffFrom(StreamInput in, K key) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    /**
+     * Implementation of ValueSerializer that delegates to another ValueSerializer.
+     * Useful for overriding single methods.
+     *
+     * @param <K> type of map keys
+     * @param <V> type of map values
+     */
+    public static class DelegateValueSerializer<K, V> implements ValueSerializer<K, V> {
+        protected final ValueSerializer<K, V> delegate;
+
+        public DelegateValueSerializer(ValueSerializer<K, V> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void writeTo(V value, StreamOutput out) throws IOException {
+            delegate.writeTo(value, out);
+        }
+
+        @Override
+        public V readFrom(StreamInput in, K key) throws IOException {
+            return delegate.readFrom(in, key);
+        }
+
+        @Override
+        public boolean supportsDiffableValues() {
+            return delegate.supportsDiffableValues();
+        }
+
+        @Override
+        public Diff<V> computeDiff(V value, V beforePart) {
+            return delegate.computeDiff(value, beforePart);
+        }
+
+        @Override
+        public void writeDiffTo(Diff<V> value, StreamOutput out) throws IOException {
+            delegate.writeDiffTo(value, out);
+        }
+
+        @Override
+        public Diff<V> readDiffFrom(StreamInput in, K key) throws IOException {
+            return delegate.readDiffFrom(in, key);
+        }
+    }
+
+    /**
+     * Implementation of the ValueSerializer that uses a prototype object for reading operations
+     *
+     * Note: this implementation is ignoring the key.
+     */
+    public static class DiffablePrototypeValueReader<K, V extends Diffable<V>> extends DelegateValueSerializer<K, V> {
+        private V proto;
+
+        public DiffablePrototypeValueReader(V proto) {
+            super(DiffableValueSerializer.getWriteOnlyInstance());
+            this.proto = proto;
+        }
+
+        @Override
+        public V readFrom(StreamInput in, K key) throws IOException {
+            return proto.readFrom(in);
+        }
+
+        @Override
+        public Diff<V> readDiffFrom(StreamInput in, K key) throws IOException {
+            return proto.readDiffFrom(in);
+        }
+    }
+
+    /**
+     * Implementation of ValueSerializer that serializes immutable HashSets
```

Comment:
I think this should read serializes immutable Sets (hash sets is an implementation detail)